File: hesabixAPI/adapters/db/models/cash_register.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import String, Integer, DateTime, ForeignKey, UniqueConstraint, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship

from adapters.db.session import Base


class CashRegister(Base):
	__tablename__ = "cash_registers"
	__table_args__ = (
		UniqueConstraint('business_id', 'code', name='uq_cash_registers_business_code'),
	)

	id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
	business_id: Mapped[int] = mapped_column(Integer, ForeignKey("businesses.id", ondelete="CASCADE"), nullable=False, index=True)

	# مشخصات
	name: Mapped[str] = mapped_column(String(255), nullable=False, index=True, comment="نام صندوق")
	code: Mapped[str | None] = mapped_column(String(50), nullable=True, index=True, comment="کد یکتا در هر کسب‌وکار (اختیاری)")
	description: Mapped[str | None] = mapped_column(String(500), nullable=True)

	# تنظیمات
	currency_id: Mapped[int] = mapped_column(Integer, ForeignKey("currencies.id", ondelete="RESTRICT"), nullable=False, index=True)
	is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="1")
	is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="0")

	# پرداخت
	payment_switch_number: Mapped[str | None] = mapped_column(String(100), nullable=True)
	payment_terminal_number: Mapped[str | None] = mapped_column(String(100), nullable=True)
	merchant_id: Mapped[str | None] = mapped_column(String(100), nullable=True)

	# زمان بندی
	created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
	updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

	# روابط
	business = relationship("Business", backref="cash_registers")
	currency = relationship("Currency", backref="cash_registers")

File: hesabixUI/hesabix_ui/lib/models/cash_register.dart
class CashRegister {
  final int? id;
  final int businessId;
  final String name;
  final String? code;
  final int currencyId;
  final bool isActive;
  final bool isDefault;
  final String? description;
  final String? paymentSwitchNumber;
  final String? paymentTerminalNumber;
  final String? merchantId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const CashRegister({
    this.id,
    required this.businessId,
    required this.name,
    this.code,
    required this.currencyId,
    this.isActive = true,
    this.isDefault = false,
    this.description,
    this.paymentSwitchNumber,
    this.paymentTerminalNumber,
    this.merchantId,
    this.createdAt,
    this.updatedAt,
  });

  factory CashRegister.fromJson(Map<String, dynamic> json) {
    return CashRegister(
      id: json['id'] as int?,
      businessId: (json['business_id'] ?? json['businessId']) as int,
      name: (json['name'] ?? '') as String,
      code: json['code'] as String?,
      currencyId: (json['currency_id'] ?? json['currencyId']) as int,
      isActive: (json['is_active'] ?? true) as bool,
      isDefault: (json['is_default'] ?? false) as bool,
      description: json['description'] as String?,
      paymentSwitchNumber: json['payment_switch_number'] as String?,
      paymentTerminalNumber: json['payment_terminal_number'] as String?,
      merchantId: json['merchant_id'] as String?,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'].toString()) : null,
      updatedAt: json['updated_at'] != null ? DateTime.tryParse(json['updated_at'].toString()) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'business_id': businessId,
      'name': name,
      'code': code,
      'currency_id': currencyId,
      'is_active': isActive,
      'is_default': isDefault,
      'description': description,
      'payment_switch_number': paymentSwitchNumber,
      'payment_terminal_number': paymentTerminalNumber,
      'merchant_id': merchantId,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}
